### PR TITLE
ResourceControl API to add standalone PNNs into the database

### LINKED ISSUE #8448: ACDC docs with empty list of files
I was wondering why the original workflow ran 106 jobs and the ACDC runs only 100 (and there are 6 lumis missing in the output). Workflow running on vocms0261

Checking the ACDC collection, the reason is that there are 6 couch docs/jobs that are missing information:
https://cmsweb-testbed.cern.ch/couchdb/acdcserver/_design/ACDC/_view/byCollectionName?key=%22amaltaro_ReDigi_AllFlag_Agent1110_Validation_180207_203747_2620%22&include_docs=true&reduce=false

e..g
```
{'doc': {'_id': 'aa85652a9be221e2bc400c6f801b95bd',
         '_rev': '1-6cd5edf31fb9d5f8fe51f67233ffa11f',
         'collection_name': 'amaltaro_ReDigi_AllFlag_Agent1110_Validation_180207_203747_2620',
         'collection_type': 'ACDC.CollectionTypes.DataCollection',
         'files': {},
         'fileset_name': '/amaltaro_ReDigi_AllFlag_Agent1110_Validation_180207_203747_2620/StepOneProc',
         'timestamp': 1518054268.63377},
 'id': 'aa85652a9be221e2bc400c6f801b95bd',
 'key': 'amaltaro_ReDigi_AllFlag_Agent1110_Validation_180207_203747_2620',
 'value': {'_id': 'aa85652a9be221e2bc400c6f801b95bd',
           '_rev': '1-6cd5edf31fb9d5f8fe51f67233ffa11f'}}
```

while it should look something like
```
{'doc': {'_id': 'aa85652a9be221e2bc400c6f801b9586',
         '_rev': '1-5f3b24bea5b279cef17c95e77590c918',
         'collection_name': 'amaltaro_ReDigi_AllFlag_Agent1110_Validation_180207_203747_2620',
         'collection_type': 'ACDC.CollectionTypes.DataCollection',
         'files': {'/store/mc/RunIIWinter15GS/BuToJpsiKV2_BMuonFilter_TuneCUEP8M1_13TeV-pythia8-evtgen/GEN-SIM/MCRUN2_71_V1_ext1-v2/80001/2EE35402-26C0-E611-AE26-F46D0450CEA0.root': {'checksums': {},
                                                                                                                                                                                       'events': 4146,
                                                                                                                                                                                       'first_event': 0,
                                                                                                                                                                                       'id': 191,
                                                                                                                                                                                       'last_event': 0,
                                                                                                                                                                                       'lfn': '/store/mc/RunIIWinter15GS/BuToJpsiKV2_BMuonFilter_TuneCUEP8M1_13TeV-pythia8-evtgen/GEN-SIM/MCRUN2_71_V1_ext1-v2/80001/2EE35402-26C0-E611-AE26-F46D0450CEA0.root',
                                                                                                                                                                                       'locations': ['T2_CH_CERN',
                                                                                                                                                                                                     'T1_US_FNAL_Disk'],
                                                                                                                                                                                       'merged': '1',
                                                                                                                                                                                       'parents': [],
                                                                                                                                                                                       'runs': [{'lumis': [198940],
                                                                                                                                                                                                 'run_number': 1}],
                                                                                                                                                                                       'size': 2337180518}},
         'fileset_name': '/amaltaro_ReDigi_AllFlag_Agent1110_Validation_180207_203747_2620/StepOneProc',
         'timestamp': 1518054268.230277},
 'id': 'aa85652a9be221e2bc400c6f801b9586',
 'key': 'amaltaro_ReDigi_AllFlag_Agent1110_Validation_180207_203747_2620',
 'value': {'_id': 'aa85652a9be221e2bc400c6f801b9586',
           '_rev': '1-5f3b24bea5b279cef17c95e77590c918'}}
```

### PR DESCRIPTION
**Pull request by amaltaro**
Fixes #8448 
Superseeds a commit or two from #8503
It also changes the PNNs added to the resource control, which will then contain the MSS endpoints again (just to make sure any valid file has a replica mapped inside the agent).

@ticoann this is what we discussed yesterday. I still need to test, and if it works Ok, then I'll do further clean up (SE stuff). 
@hufnagel maybe there was a good reason we didn't add MSS to resource control anymore?

### CODE REVIEW

**Comment by hufnagel** [comment]
The reason we don't track MSS PNN in WMAgent is simple (and still valid IMO). We don't process files from tape and MSS PNN never map to any PSN, since no PSN has direct access to data stored in MSS PNN.

**Comment by amaltaro** [comment]
I see. Still, PNN is a data location attribute and I think we could treat a file location regardless whether it's possible to process it (it has a PSN associated to) or not.

All of this is basically to solve one problem, which is, the agent might acquire a block that is located at MSS/Buffer only (with TrustSitelist flag), which would cause all those files not to have any valid location in wmbs_file_details. More info in the GH issue linked above.

Provided it doesn't cause any bad side-effect, I think it's then fine to have these changes in :)

**Comment by hufnagel** [comment]
Well, if you acquire data that is only present at MSS PNN and run a workflow on them with TrustSitelist set, all these jobs will fail since MSS PNN are not part of the AAA data federation and you can't read remotely from MSS PNN.

**Comment by hufnagel** [comment]
I am not saying you can't include them. If you do, they are going to be special though and need special treatment. Like workflows with data only at MSS PNN are workflow with data on disk, TrustSitelist set or not.

**Comment by amaltaro** [comment]
> Well, if you acquire data that is only present at MSS PNN and run a workflow on them with TrustSitelist set, all these jobs will fail since MSS PNN are not part of the AAA data federation and you can't read remotely from MSS PNN.

I know, but the acdc collections will at least have the correct file location (which is MSS, not an empty list)

**Comment by amaltaro** [comment]
Ok, my quick test succeeded and I see the correct info in the ACDC docs, e.g.:
```
...{"lfn":"/store/mc/RunIIWinter15GS/BuToJpsiKV2_BMuonFilter_TuneCUEP8M1_13TeV-pythia8-evtgen/GEN-SIM/MCRUN2_71_V1_ext1-v2/80001/68C53F10-BBC0-E611-9DDB-0CC47A075DBC.root","locations":["T1_UK_RAL_MSS"]...
```

I think this change will be cleaner, so I'll bring the other ErrorHandler improvements over to this PR.

**Comment by amaltaro** [comment]
The third commmit (actually the fourth one) has some SiteDB API cleanup. I removed everything related to storage element, however I'm not sure we want to remove those apis, since SiteDB still has the storage element endpoint as well(?)

**Comment by ticoann** [comment]
test this please

File: src/python/WMCore/WMBS/Oracle/Locations/AddPNNs.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python
+"""
+Oracle implementation of AddPNNs
+"""
+
+from WMCore.WMBS.MySQL.Locations.AddPNNs import AddPNNs as AddPNNsMySQL
+
+
+class AddPNNs(AddPNNsMySQL):
+    pass
```
Comment (ticoann) [inline src/python/WMCore/WMBS/Oracle/Locations/AddPNNs.py]: INSERT IGNORE doesn't work in Oracle.  should do something like below (not tested)

"""
INSERT /*+ IGNORE_ROW_ON_DUPKEY_INDEX (wmbs_pnns (id, pnn)) */  INTO wmbs_pnns (id, pnn)
 SELECT wmbs_pnns_seq.nextval, :pnn
"""

File: src/python/WMCore/WMBS/CreateWMBSBase.py [hunk 2]
```diff
@@ -494,6 +494,9 @@ def __init__(self, logger=None, dbi=None, params=None):
         self.constraints["01_idx_wmbs_file_checksums"] = \
             """CREATE INDEX idx_wmbs_file_checksums_file ON wmbs_file_checksums(fileid) %s""" % tablespaceIndex
 
+        self.constraints["01_idx_wmbs_pnns"] = \
+            """CREATE INDEX idx_wmbs_pnns_pnn ON wmbs_pnns(pnn) %s""" % tablespaceIndex
```
Comment (amaltaro) [inline src/python/WMCore/WMBS/CreateWMBSBase.py]: Apparently I forgot creating this index when this code was put in place, in the previous release.

**Comment by amaltaro** [comment]
test this please

File: src/python/WMCore/WMBS/CreateWMBSBase.py (hunk 2 above)
Comment (amaltaro) [reply src/python/WMCore/WMBS/CreateWMBSBase.py]: Nevermind, the primary key (and apparently unique) constraint already creates an index for that column. Removing it

File: src/python/WMCore/WMBS/MySQL/Jobs/LoadForErrorHandler.py [hunk 3]
```diff
@@ -19,25 +19,16 @@ class LoadForErrorHandler(DBFormatter):
...
-               INNER JOIN wmbs_users ON ww.owner = wmbs_users.id
-               LEFT OUTER JOIN wmbs_location ON
-                 wmbs_job.location = wmbs_location.id
-               LEFT OUTER JOIN wmbs_job_state ON
-                 wmbs_job.state = wmbs_job_state.id
              WHERE wmbs_job.id = :jobid"""
 
-    fileSQL = """SELECT wfd.id, wfd.lfn, wfd.filesize size, wfd.events, wfd.first_event,
-                   wfd.merged, wja.job jobid, wpnn.pnn
+    fileSQL = """SELECT wfd.id, wfd.lfn, wfd.filesize AS size, wfd.events, wfd.first_event,
```
Comment (amaltaro) [inline src/python/WMCore/WMBS/MySQL/Jobs/LoadForErrorHandler.py]: I thought I could use the same syntax for oracle, but it actually complains about "ORA-00923: FROM keyword not found where expected", I'm rolling back that change on the oracle class.

**Comment by ticoann** [comment]
test this please

**Comment by amaltaro** [comment]
@ticoann can you please review it? I think we're ready to backport these patches, I'l just wait for your review here first.

**Comment by ticoann** [comment]
test this please

**Comment by ticoann** [comment]
@amaltaro, Alan patch looks good to me. Could you squash the commit?

**Comment by amaltaro** [comment]
Done! I left the ACDC speed up in a separate commit because it's orthogonal from each other.
PS.: Don't ask me why the unit tests commit is in the middle instead of in the bottom (in my git log it shows at the very top...)

**Comment by ticoann** [comment]
Thanks Alan. :-)

**Comment by ticoann** [comment]
test this please

**Comment by ticoann** [comment]
test this please

**Comment by ticoann** [comment]
Yes, Alan. I was just wondering why jenkins report is keep failing even though all the unittest is successful.

**Comment by amaltaro** [comment]
Because it failed the pylint checks
"Pylint check: failed "

one (or some) of those files have warning level issues.

**Comment by ticoann** [comment]
test this please

**Comment by ticoann** [comment]
But it still reports that I only see default check fails.

**Comment by ticoann** [comment]
Thanks Eric for investigating.

**Pull request merged**